include/swift-c/DependencyScan/DependencyScan.h

```diff
@@ -0,0 +1,228 @@
+ds_get_swift_textual_detail_bridging_module_dependencies(
+    ds_module_details_t details);
+
+DEPSCAN_PUBLIC ds_string_set_t *
+ds_get_swift_textual_detail_command_line(ds_module_details_t details);
+
+DEPSCAN_PUBLIC ds_string_set_t *
+ds_get_swift_textual_detail_extra_pcm_args(ds_module_details_t details);
+
+DEPSCAN_PUBLIC ds_string_t
+ds_get_swift_textual_detail_context_hash(ds_module_details_t details);
+
+DEPSCAN_PUBLIC bool
+ds_get_swift_textual_detail_is_framework(ds_module_details_t details);
+
+//=== Swift Binary Module Details query APIs ------------------------------===//
+
+DEPSCAN_PUBLIC ds_string_t
+ds_get_swift_binary_detail_compiled_module_path(ds_module_details_t details);
+
+DEPSCAN_PUBLIC ds_string_t
+ds_get_swift_binary_detail_module_doc_path(ds_module_details_t details);
+
+DEPSCAN_PUBLIC ds_string_t ds_get_swift_binary_detail_module_source_info_path(
+    ds_module_details_t *details);
+
+//=== Swift Placeholder Module Details query APIs -------------------------===//
+
+DEPSCAN_PUBLIC ds_string_t ds_get_swift_placeholder_detail_compiled_module_path(
+    ds_module_details_t details);
+
+DEPSCAN_PUBLIC ds_string_t
+ds_get_swift_placeholder_detail_module_doc_path(ds_module_details_t details);
+
+DEPSCAN_PUBLIC ds_string_t
+ds_get_swift_placeholder_detail_module_source_info_path(
+    ds_module_details_t details);
+
+//=== Clang Module Details query APIs -------------------------------------===//
+
+DEPSCAN_PUBLIC ds_string_t
+ds_get_clang_detail_module_map_path(ds_module_details_t details);
+
+DEPSCAN_PUBLIC ds_string_t
+ds_get_clang_detail_context_hash(ds_module_details_t details);
+
+DEPSCAN_PUBLIC ds_string_set_t *
+ds_get_clang_detail_command_line(ds_module_details_t details);
+
+//=== Cleanup Functions ---------------------------------------------------===//
+
+DEPSCAN_PUBLIC void
+ds_dependency_info_details_dispose(ds_module_details_t details);
+
+DEPSCAN_PUBLIC void ds_dependency_info_dispose(ds_dependency_info_t *info);
+
+DEPSCAN_PUBLIC void ds_dependency_set_dispose(ds_dependency_set_t *set);
+
+DEPSCAN_PUBLIC void
+ds_dependency_result_dispose(ds_dependency_result_t *result);
+
+DEPSCAN_PUBLIC void ds_prescan_result_dispose(ds_prescan_result_t *result);
+
+DEPSCAN_PUBLIC void ds_batch_scan_entry_dispose(ds_batch_scan_entry_t *entry);
+
+DEPSCAN_PUBLIC void ds_batch_scan_input_dispose(ds_batch_scan_input_t *input);
+
+DEPSCAN_PUBLIC void
+ds_batch_scan_result_dispose(ds_batch_scan_result_t *result);
+
+//=== Scanner Functions ---------------------------------------------------===//
+
+/// Container of the configuration state and shared cache for dependency
+/// scanning.
+typedef void *ds_scanner_t;
+
+DEPSCAN_PUBLIC ds_scanner_t ds_scanner_create(void);
+
+DEPSCAN_PUBLIC void ds_scanner_dispose(ds_scanner_t);
+
+DEPSCAN_PUBLIC ds_dependency_result_t *
+ds_scan_dependencies(ds_scanner_t *scanner, const char *working_directory,
+                     int argc, const char *const *argv);
```

Comment:
How confident are you that you never need more parameters to this function (and the two that follow)?  To allow addition, you could create an opaque "request" with setter functions for the things you need.  Or as a middle ground, you could expose an extra `void *` parameter that should be set to `nullptr` for now, but it leaves you room to add extended parameters in the future.

Reply:
Added an opaque `invocation` context type with setters/getters for its fields and a creation method. Thank you.